### DISCUSSION TITLE
Update dependencies in the sub-module build.sbt

### DISCUSSION THREAD

**Discussion by yadavan88** [scala-steward-org/scala-steward#3151] [Q&A] [answered]
Hi there,
Sorry, I am here again with one more question. 
I have a project with this structure
```
root-module
         | --- build.sbt 
         | --- group-modules
                         | ----- build.sbt //just to aggregate inner modules
                         | ----- sub1
                                      | ----- build.sbt 
```

There are a few dependencies defined inside the `sub1/build.sbt`. However, these deps are not updated using scala-steward. 

Do I need to do something to manage these deps?

Thanks in advance
Yadu

**Comment by yadavan88** [comment]
Just an update. Due to a different reason, I switched to the scala-steward github action. With that, the scala-steward PRs are created even for inner modules. When I was using the public instance, then it was not happening.

**Comment by fthomas** [comment, marked as answer]
> Do I need to do something to manage these deps?

Yes, you need to set the `buildRoots` in your `.scala-steward.conf` to also update projects in subdirectories. Here is the relevant documentation from https://github.com/scala-steward-org/scala-steward/blob/main/docs/repo-specific-configuration.md:

> It is possible to have multiple scala projects in a single repository. In that case the folders containing the projects (build.sbt folders)
> are specified using the buildRoots property. Note that the paths used there are relative and if the repo directory itself also contains a build.sbt the dot can be used to specify it.
> Default: ["."]
> buildRoots = [ ".", "subfolder/projectA" ]

**Reply by yadavan88** [reply]
Hi @fthomas 
Sorry to bother again. It seems the PRs are not generated. I can't set the build root in this particular case since the inner sub-modules are not independent projects. 
This is for a tutorial project I am helping to maintain. There are a bunch of play samples, each one of them are grouped in a directory called as play-scala. However, they are not aggregated specifically. 

Would it be possible for you to have a look at this and suggest what to do in this case? 
Main Repo: https://github.com/Baeldung/scala-tutorials
Sub-Module which just aggregate play samples: https://github.com/Baeldung/scala-tutorials/tree/master/play-scala

What I want is to automatically update the dependencies in each sub-modules within play-scala directory.

Thanks in advance

**Reply by fthomas** [reply]
@yadavan88 It seems to me that Scala Steward cannot handle repositories with such layout. It supports only sbt builds where calling `sbt libraryDependencies` lists all dependencies that should be updated. If the submodules are not independent so that `buildRoots` can't be used and the root project doesn't aggregate all subprojects, there is no way for Scala Steward to extract dependencies from the build.

**Reply by yadavan88** [reply]
Hi @fthomas Thank you for quick reply and clarification. I really appreciate that. :) 
So if I want to manage the dependencies, I need to move the build definitions from inner dir to the main build.sbt, right?

**Reply by fthomas** [reply]
Maybe you could just aggregate the subprojects in the top-level project with
```
lazy val root = project
  .in(file("."))
  .aggregate(cats_effects, doobie, scalatra, ...)
```

If an invocation of `sbt libraryDependencies` in the top-level directory lists all dependencies of the subprojects, Scala Steward should be able to update them.

**Reply by yadavan88** [reply]
I think I will try this approach. This was once tried, but later decided not to do since this is a tutorial repo and there are too many sub modules. If someone misses one from aggregation, it will be missed out from compile and tests. But if we didn't use explicit aggregation, sbt automatically aggregates all project definitions in the build.sbt. 
But now we need to evaluate what is more important here. Thank you very much for your help.
